doc/fips/SONiC-OpenSSL-FIPS-140-3.md

```diff
@@ -0,0 +1,155 @@
+dynamic_path = /usr/lib/x86_64-linux-gnu/libsymcryptengine.so
+default_algorithms = ALL
+```
+
+### OpenSSL configuration enhancement
+When fips=1 is set in /proc/cmdline, the OpenSSL default config file is changed to "/usr/lib/ssl/openssl-fips.cnf", otherwise, the config file "/usr/lib/ssl/openssl.cnf" is used.
+
+### SymCrypt OpenSSL Engine debian package
+Provide SymCrypt OpenSSL debian package.
+Package name: symcrypt-openssl
+Current version: 0.1
+
+Package file name example: symcrypt-openssl_0.1_amd64.deb
+Files in the packages:
+```
+/usr/lib/ssl/openssl.cnf
+/usr/lib/ssl/openssl-fips.cnf
+/usr/lib/x86_64-linux-gnu/libsymcrypt.so
+/usr/lib/x86_64-linux-gnu/libsymcryptengine.so
+```
+
+## Kerberos Cryptographic Module
+Kerberos will use the builtin cryptographic module by default, but it allows to change the build option to use OpenSSl, see [MIT Kerberos features](https://web.mit.edu/kerberos/krb5-1.13/doc/mitK5features.html). SONiC will change the build option to use OpenSSL instead of the builtin one. It is not configurable to use the Kerberos builtin cryptographic module when OpenSSL used.
+
+## Golang Cryptographic Module
+Golang has its own cryptographic module (see [crypto](https://github.com/golang/go/tree/master/src/crypto)) without FIPS supports. There are some branches with branch name starting with "dev.boringcrypto" (see [golang branches](https://github.com/golang/go/branches/all?query=dev.boringcrypto)), changing the Golang cryptographic APIs' referenece to use [BoringSSL](https://github.com/google/boringssl). Although BoringSSL is an open source project, but it used by Google only, not intened for general use.
+
+To support FIPS for Golang, RedHat offers an alternative solution (see [here](https://developers.redhat.com/blog/2019/06/24/go-and-fips-140-2-on-red-hat-enterprise-linux)), it builds on top of the Golang's dev.bringcrypt branches, has ability to call into OpenSSL, not BoringSSL. SONiC can reuse the RedHat sulotion, one difference is that RedHat supports FIPS for OpenSSL directly, SONiC uses OpenSSL Engine.
+
+How OpenSSL Engine works in Golang?
+![Golang API](images/golang-api.png)
+
+When FIPS enabled, both of the BoringSSL Enable Option and the SymCrypt Enabled option will be set.
+
+
+## Application Impact
+Some of functions of a application might be broken when using the cryptographic algorithms that are not FIPS compliant. It is relied on the tests of the applications to detect all the impact functions.
+For OpenSSH, Centos provides a [patch](https://git.centos.org/rpms/openssh/raw/c8/f/SOURCES/openssh-7.7p1-fips.patch) which is compiant with FIPS 140-2. We can apply the patch and verify if it can pass all the OpenSSH test cases when FIPS enabled.
+
+## SONiC FIPS Configuration
+### Enable FIPS on system level
+Add the Linux System parameter fips=1, in grub config, one of implemetation as below:
+
+cat /etc/grub.d/99-fips.cfg
+```
+GRUB_CMDLINE_LINUX_DEFAULT="$GRUB_CMDLINE_LINUX_DEFAULT fips=1"
+```
```

Comment:
We will have multiple copies of OpenSSL in host and dockers (e.g. in SNMP docker for SNMPv3, in Telemetry docker - for gNMI etc.) Each one should be configured to link/use the SymCrypt Engine. How will the global setting on running a kernel shared by all dockers impact different OpenSSL copies used in different dockers ?

Reply:
@eddyk-nvidia , not necessary to copy to different dockers, by default, the /proc/cmdline in all the docker containers are the same the the one in host system.

Reply:
So, setting "fips=1" to the kernel cmdline we will have SynCrypt engine be used by OpenSSL in all dockers, right ?

Reply:
@eddyk-nvidia , yes, the config is available in all dockers, not additional setting required. But it is required to install the engine in all dockers (we can install it in docker base image).